crates/driver/src/domain/competition/solution/encoding.rs

```diff
@@ -7,3 +22,330 @@ pub enum Strategy {
     /// Use logic from this module for encoding
     Domain,
 }
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("invalid interaction: {0:?}")]
+    InvalidInteractionExecution(competition::solution::interaction::Liquidity),
+    #[error("invalid clearing price: {0:?}")]
+    InvalidClearingPrice(eth::TokenAddress),
+    #[error(transparent)]
+    Math(#[from] Math),
+}
+
+pub fn tx(
+    auction_id: competition::auction::Id,
+    solution: &super::Solution,
+    contract: &contracts::GPv2Settlement,
+    approvals: impl Iterator<Item = eth::allowance::Approval>,
+    internalization: settlement::Internalization,
+) -> Result<eth::Tx, Error> {
+    let mut tokens = Vec::new();
+    let mut clearing_prices = Vec::new();
+    let mut trades: Vec<Trade> = Vec::new();
+    let mut pre_interactions = Vec::new();
+    let mut interactions = Vec::new();
+    let mut post_interactions = Vec::new();
+
+    // Encode uniform clearing price vector
+    for (token, price) in solution.prices.clone() {
+        tokens.push(token.0.into());
+        clearing_prices.push(price);
+    }
+
+    // Encode trades with custom clearing prices
+    for trade in solution.trades() {
+        match trade {
+            super::Trade::Fulfillment(trade) => {
+                tokens.push(trade.order().sell.token.0.into());
+                tokens.push(trade.order().buy.token.0.into());
+
+                let uniform_prices = ClearingPrices {
+                    sell: solution
+                        .clearing_price(trade.order().sell.token)
+                        .ok_or(Error::InvalidClearingPrice(trade.order().sell.token))?,
+                    buy: solution
+                        .clearing_price(trade.order().buy.token)
+                        .ok_or(Error::InvalidClearingPrice(trade.order().buy.token))?,
+                };
+                let custom_prices = trade.custom_prices(&uniform_prices)?;
+                clearing_prices.push(custom_prices.sell);
+                clearing_prices.push(custom_prices.buy);
+
+                trades.push(Trade {
+                    sell_token_index: (tokens.len() - 2).into(),
+                    buy_token_index: (tokens.len() - 1).into(),
+                    receiver: trade.order().receiver.unwrap_or_default().into(),
+                    sell_amount: trade.order().sell.amount.into(),
+                    buy_amount: trade.order().buy.amount.into(),
+                    valid_to: trade.order().valid_to.0.into(),
```

Comment:
We do not need to access its internal type

```suggestion
                    valid_to: trade.order().valid_to.into(),
```